Add Fill-in-the-middle training objective example - PyTorch

tanaymeh opened the pull request:
# What does this PR do?
This PR adds a Fill-in-the-middle training objective example to 🤗 transformers.

FIM objective was proposed in [Efficient Training of Language Models to Fill in the Middle](https://arxiv.org/abs/2207.14255). They showed that autoregressive language models can learn to infill text after applying a straightforward transformation to the dataset, which simply moves a span of text from the middle of a document to its end.

As discussed in #27059 

## Who can review?
@sayakpaul @ArthurZucker

tanaymeh committed:
add: initial script to train clm fim

tanaymeh commented:
This PR is currently in draft mode until I can get a GPU VM (shouldn't take more than a week at max) to test to if a model is training all fine using this script.

tanaymeh committed:
fix: if training model from scratch, new tokens will be added and embeddings resized

tanaymeh committed:
fix: fixed attention_mask errors when generating FIM data

tanaymeh committed:
fix: file formatted using black

tanaymeh committed:
add: run_fim_no_trainer.py and fixed some comments in run_fim.py

tanaymeh committed:
add: added fim examples to the README.md and ran code fixup

tanaymeh committed:
fix: little bug in both fim training scripts

tanaymeh commented:
@sayakpaul @ArthurZucker This PR is ready for review. I have added two scripts: `run_fim.py` and `run_fim_no_trainer.py`. Both have almost the same code structure as `run_clm.py` and `run_clm_no_trainer.py` respectively.

I have adapted the FIM dataset transformation from the [SantaCoder repository](https://github.com/loubnabnl/santacoder-finetuning/blob/main/fim.py#L22C13-L83) with changes to make it work with dataset's `.map()` function.

Training in both scripts works but I have yet to train a proper model on a GPU to compare the performance. Let me know if I need to make any changes to the script.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27464). All of your documentation changes will be reflected on that endpoint.

tanaymeh committed:
fix: remove comment from notebook and added a note on fim related params

tanaymeh committed:
fix: minor typo in README

tanaymeh commented:
@ArthurZucker @pacman100 Are there any issues in this PR? Kindly let me know!

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,799 @@
+            )
+        },
+    )
+    use_auth_token: bool = field(
+        default=None,
+        metadata={
+            "help": "The `use_auth_token` argument is deprecated and will be removed in v4.34. Please use `token` instead."
+        },
```
let's not pass this as it's deprecated
```suggestion
    token: str = field(default=None,metadata={"help": "The `token` blah blah"},
```

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,799 @@
+        default="<fim_suffix>",
+        metadata={"help": ("Fill-in-Middle Suffix token. Defaults to '<fim_suffix>'.")},
+    )
+    fim_pad_token: Optional[str] = field(
+        default="<fim_pad>",
+        metadata={
+            "help": (
+                "Fill-in-Middle Pad token. Used only when 'truncate_or_pad' is set to True. "
+                "Defaults to '<fim_pad>'."
+            )
```
this can just be the pad token.

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,799 @@
+            if model_args.torch_dtype in ["auto", None]
+            else getattr(torch, model_args.torch_dtype)
+        )
+        model = AutoModelForCausalLM.from_pretrained(
+            model_args.model_name_or_path,
+            from_tf=bool(".ckpt" in model_args.model_name_or_path),
+            config=config,
+            cache_dir=model_args.cache_dir,
+            revision=model_args.model_revision,
+            token=model_args.token,
+            trust_remote_code=model_args.trust_remote_code,
+            torch_dtype=torch_dtype,
+            low_cpu_mem_usage=model_args.low_cpu_mem_usage,
+        )
+
+    else:
+        model = AutoModelForCausalLM.from_config(config, trust_remote_code=model_args.trust_remote_code)
+        n_params = sum({p.data_ptr(): p.numel() for p in model.parameters()}.values())
+        logger.info(f"Training new model from scratch - Total size={n_params/2**20:.2f}M params")
```
we should support `attn_implementation` and wait for it to be merged. 
I think we should try to incorporate as many changes as possible from latest release. Like gradient checkpointing argument, the pad to multiple of when resizing the token embedding to optimize for the architecture

ArthurZucker replied:
Next week 😉 q

ArthurZucker replied:
Release done!

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,799 @@
+        special_tokens.append(data_args.fim_pad_token)
+
+    tokenizer.add_tokens(special_tokens)
+    model.resize_token_embeddings(len(tokenizer))
```
resizing should be handled like this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html

ArthurZucker replied:
No what I mean is using the average of the already available tokens not necessarly wte. You can always get the embeding with `get_input_embeddings` and `set_input_embeddings`

tanaymeh replied:
@ArthurZucker I am doing this for adding new tokens' embeddings: they are set to the average of the existing embeddings. Is this a good approach?

```python
model.resize_token_embeddings(len(model.get_input_embeddings().weight) + len(new_tokens))

# Get the updated embedding layer
embeddings = model.get_input_embeddings()

# Calculate the average of the existing embeddings
avg_embedding = embeddings.weight[:-len(new_tokens)].mean(dim=0)
new_embeddings = embeddings.weight.clone()

# Set the new tokens' embeddings to this average
new_embeddings[-len(new_tokens):] = avg_embedding

# Update the model's embedding layer
embeddings.weight = torch.nn.Parameter(new_embeddings)
model.set_input_embeddings(embeddings)
```

ArthurZucker replied:
It is, but this will make the embeddings equal, you need to sample from a uniform distribution centered around the mean of the embeddings! But all good otherwise

tanaymeh replied:
Thanks, @ArthurZucker , I have a doubt,
When using `pad_to_mulitple_of` argument, should the tokens `len(new_tokens)` actually reflect all the new tokens added, including the ones added due to padding?

For example, if `4` FIM tokens and extra `3` tokens were added to GPT2 to make the embedding size `50264` (multiple of 8), should the embeddings' sampling be done all 7 tokens or just the original `4` FIM tokens only?

ArthurZucker replied:
I think the documentation should explain this but the idea is you just say you want len(tokenizer), the pad_to_multiple of has nothing to do with this. You should set it according to the device you are using

tanaymeh replied:
I see, thanks for clarifying this!

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,799 @@
+    model.resize_token_embeddings(len(tokenizer))
+    print("Added special tokens!")
+
+    # We resize the embeddings only when necessary to avoid index errors. If you are creating a model from scratch
+    # on a small vocab and want a smaller embedding size, remove this test.
+    embedding_size = model.get_input_embeddings().weight.shape[0]
+    if len(tokenizer) > embedding_size:
+        model.resize_token_embeddings(len(tokenizer))
```
duplicated?

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,799 @@
+            return metric.compute(predictions=preds, references=labels)
+
+    # Initialize our Trainer
+    trainer = Trainer(
```
trainer or sft trainers could be a nice to support as well

ArthurZucker reviewed (commented):
thanks for this! Could you also add maybe and example log from the run to the readme>?

tanaymeh commented on examples/pytorch/language-modeling/README.md:
````diff
@@ -73,6 +73,55 @@ python run_clm_no_trainer.py \
+```
+
+<!-- **Few things to consider when running `run_fim.py` and `run_fim_no_trainer.py`**:
+1. If you are -->
````
I have removed this in the latest commit!

tanaymeh committed:
add: suggested minor changes to README and run_fim.py

tanaymeh commented:
> thanks for this! Could you also add maybe and example log from the run to the readme>?

@ArthurZucker, Since the training log is almost entirely the same as `run_clm.py` (the only difference is the dataset transformation for adding FIM tokens), should I instead add the parts of the log concerned with the training part only?

ArthurZucker commented:
Sure 🤗

tanaymeh committed:
add: gradient_accumulation_steps and gradient_checkpointing args

tanaymeh committed:
add: improved model embedding resizing

tanaymeh commented:
@ArthurZucker I added `MultivariateNormal` distribution-based sampling of new token embeddings as was in the blog.

I have a question, should this normalisation of new embeddings also include the case when we are using `pad_to_mulitple_of`?

For example, if `pad_to_multiple_of` increases the embedding size by 10, should 10 be now added to the existing new 4 tokens we are adding, requiring us to sample 14 new token embeddings and add them to the embedding layer?

tanaymeh committed:
add: pad_to_multiple_of and attn_implementation params

tanaymeh commented:
@ArthurZucker I have added the `pad_to_multiple_of` argument based on the device (for GPU it will be 8 and for TPUs, it will be 128), you can see it [here](https://github.com/tanaymeh/transformers/blob/98a5976cc555696c2ae0e4eb891b9e77d989c291/examples/pytorch/language-modeling/run_fim.py#L521-L533).

I have added `attn_implementation` support in the models ([here](https://github.com/tanaymeh/transformers/blob/98a5976cc555696c2ae0e4eb891b9e77d989c291/examples/pytorch/language-modeling/run_fim.py#L504)) and the model's embedding resizing changes as you requested ([here](https://github.com/tanaymeh/transformers/blob/98a5976cc555696c2ae0e4eb891b9e77d989c291/examples/pytorch/language-modeling/run_fim.py#L529-L560)).

Please let me know if these are the correct changes or do I need to make any amends!

ArthurZucker commented:
Hey! Sorry for the delay, it's a big PR I'll review it later this week 🤗

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,832 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,832 @@
+Here is the full list of checkpoints on the hub that can be fine-tuned by this script:
+https://huggingface.co/models?filter=text-generation
+"""
+# You can also adapt this script on your own causal language modeling task. Pointers for this are left as comments.
```
```suggestion
# You should adapt this script on your own causal language modeling task. Pointers for this are left as comments.
```

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,832 @@
+            )
+        },
+    )
+    pad_to_multiple_of: bool = field(
```
nice

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,832 @@
+        },
+    )
+    attn_implementation: Optional[str] = field(
+        default=None, metadata={"help": ("The attention implementation to use. ")}
```
```suggestion
        default="sdpa", metadata={"help": ("The attention implementation to use. ")}
```
to make sure it's explicit

ArthurZucker commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,832 @@
+    original_embeddings = model.get_input_embeddings().weight
+    model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=pad_factor)
+
+    # Sample the embeddings for the new tokens from a multivariate normal distribution
+    # We do this so that the new embeddings are close to the original embeddings and not necessarily zero
+    # More on this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html
```
nice 🤗 !

ArthurZucker commented on examples/pytorch/language-modeling/run_fim_no_trainer.py:
```diff
@@ -0,0 +1,850 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

ArthurZucker reviewed (commented):
Thanks for the great work here! Comments are good and should help. he way new tokens are dealt is also very nice. LGTM! 🤗

ArthurZucker reviewed (approved):
Let's fix the few nits and good to merge

tanaymeh committed:
add: requested minor changes

tanaymeh commented:
@ArthurZucker Thanks for your review! I have added your suggested changes and also made sure all CIs are now green.

Just waiting for @pacman100's review to get this merged!

pacman100 commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,831 @@
+        fim_transform_ids = [fim_transform(ids) for ids in examples["input_ids"]]
+        examples["input_ids"] = fim_transform_ids
+        examples["labels"] = fim_transform_ids
+        # If your application requires custom attention mask, please adjust this function's below line
+        # since FIM transformation increases the number of tokens in input_ids and labels
```
```suggestion
        # If your application requires custom attention mask, please adjust this function's below line.
        # Since FIM transformation increases the number of tokens in input_ids and labels
```

pacman100 reviewed (commented):
Thank you @tanaymeh for the example scripts on FIM Task, overall LGTM! 🔥🚀✨

Just that the embedding resizing logic with init using Gaussian dist of original embeddings would fail when using DeepSpeed ZeRO3 as the weights are sharded and you would need to use `deepspeed.zero.GatheredParameters` for gathering the embedding layer weights before modifying them.

tanaymeh commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,831 @@
+        dim=0,
+    )
+
+    # Get the updated embedding layer and make a copy of it's weights
+    embeddings = model.get_input_embeddings()
+    new_embeddings = embeddings.weight.clone()
+
+    # Set the new tokens' embeddings to the newly sampled embeddings
+    new_embeddings[-len(special_tokens) :] = new_token_embeddings
+
+    # Update the model's embeddings with the new embeddings
+    embeddings.weight = torch.nn.Parameter(new_embeddings)
+    model.set_input_embeddings(embeddings)
```
Thanks for the review @pacman100!

I had a doubt, shouldn't I be using `deepspeed.zero.GatheredParameters` for that? It will look something like this:

```python
if is_deepspeed_zero3_enabled():
    import deepspeed
    with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=None):
            new_embeddings = embeddings.weight.clone()

            # Set the new tokens' embeddings to the newly sampled embeddings
            new_embeddings[-len(special_tokens) :] = new_token_embeddings
        
            # Update the model's embeddings with the new embeddings
            embeddings.weight = torch.nn.Parameter(new_embeddings)
            model.set_input_embeddings(embeddings)
```

pacman100 replied:
Yes, I meant this, updated the original comment.

ArthurZucker commented:
@tanaymeh just waiting to make sure the deepspeed thing works and we should be able to merge

tanaymeh committed:
add: deepspeed zero compatibility

tanaymeh commented:
@ArthurZucker @pacman100 Thanks for the reviews and sorry for the delay, I was busy with my coursework.

I have added the suggested changes and the deepspeed compatibility to both trainer and non-trainer scripts. Could you please review and merge?

TIA!

pacman100 commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,843 @@
+
+    # Add the new tokens to the tokenizer
+    tokenizer.add_tokens(special_tokens)
+    original_embeddings = model.get_input_embeddings()
```
Even `roiginal_embeddings` needs to be gathered when using DS ZeRO3 with zero init enabled.

pacman100 commented on examples/pytorch/language-modeling/run_fim.py:
```diff
@@ -0,0 +1,843 @@
+    if is_deepspeed_zero3_enabled():
+        import deepspeed
+
+        with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=None):
+            # Set the new tokens' embeddings to the newly sampled embeddings
+            new_embeddings = embeddings.weight.clone()
+            new_embeddings[-len(special_tokens) :] = new_token_embeddings
+
+            # Update the model's embeddings with the new embeddings
+            embeddings.weight = torch.nn.Parameter(new_embeddings)
+            model.set_input_embeddings(embeddings)
```
```suggestion
        with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=0):
            # Set the new tokens' embeddings to the newly sampled embeddings
            embeddings.weight.data[-len(special_tokens) :] = new_token_embeddings
```

pacman100 replied:
as the embeddings are being modified, the modifier rank must be given. Also, I think it is enough to just change the data.

tanaymeh replied:
Thanks for the feedback @pacman100 

Is this below a good approach? I believe I will have to call GatheredParameters twice, once for original_embeddings and once for embeddings.

```python
    original_embeddings = model.get_input_embeddings()

    if is_deepspeed_zero3_enabled():
        import deepspeed
        with deepspeed.zero.GatheredParameters(original_embeddings.weight, modifier_rank=0):
            # Get the pre-expansion embeddings of the model and resize the embedding layer
            model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=pad_factor)
            embeddings = model.get_input_embeddings()

            # Sample the embeddings for the new tokens from a multivariate normal distribution
            # We do this so that the new embeddings are close to the original embeddings and not necessarily zero
            # More on this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html
            mean = original_embeddings.mean(dim=0)
            n = original_embeddings.size()[0]
            sigma = ((original_embeddings - mean).T @ (original_embeddings - mean)) / n
            dist = torch.distributions.multivariate_normal.MultivariateNormal(
                mean,
                covariance_matrix=1e-5 * sigma,
            )
            new_token_embeddings = torch.stack(
                tuple((dist.sample() for _ in range(len(special_tokens)))),
                dim=0,
            )
    else:
        original_embeddings = model.get_input_embeddings()
        # Get the pre-expansion embeddings of the model and resize the embedding layer
        model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=pad_factor)
        embeddings = model.get_input_embeddings()

        # Sample the embeddings for the new tokens from a multivariate normal distribution
        # We do this so that the new embeddings are close to the original embeddings and not necessarily zero
        # More on this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html
        mean = original_embeddings.mean(dim=0)
        n = original_embeddings.size()[0]
        sigma = ((original_embeddings - mean).T @ (original_embeddings - mean)) / n
        dist = torch.distributions.multivariate_normal.MultivariateNormal(
            mean,
            covariance_matrix=1e-5 * sigma,
        )
        new_token_embeddings = torch.stack(
            tuple((dist.sample() for _ in range(len(special_tokens)))),
            dim=0,
        )

    if is_deepspeed_zero3_enabled():
        import deepspeed
        with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=0):
            # Set the new tokens' embeddings to the newly sampled embeddings
            embeddings.weight.data[-len(special_tokens) :] = new_token_embeddings
    else:
        # Set the new tokens' embeddings to the newly sampled embeddings
        embeddings.weight.data[-len(special_tokens) :] = new_token_embeddings
    
    # Update the model's embeddings with the new embeddings
    model.set_input_embeddings(embeddings)
```

pacman100 replied:
Hello @tanaymeh, the above changes are correct, Thank you!

pacman100 reviewed (commented):
Hello @tanaymeh, wrt DeepSpeed few changes are still required. Please see the comment

ArthurZucker commented:
I already approved, I am just waiting for @pacman100's approval here! 🤗

pacman100 reviewed (approved):
Approving the PR, good to merge once the changes in the comment https://github.com/huggingface/transformers/pull/27464/files#r1505771256 are committed. Thank you @tanaymeh for all the work on adding examples for finetuning/pretraining code infilling models! 🔥🚀✨

tanaymeh committed:
add: resize embeddings layer with zero3 support for fim model initialization

tanaymeh commented:
Thanks for approving the PR @ArthurZucker and @pacman100!

I have pushed all the changes and the PR should be ready to be merged 🤗

amyeroberts commented:
@tanaymeh Thanks for working on this! I'll merge as I can see the requested changes are added and I'm on repo watch this week :)
